replace.1
.TH replace 1 ""
.SH NAME

replace \- Replace strings in text or binary files in a sane fashion

.SH SYNOPSIS

.B replace
.RB [ -? ]
.RB [ -A ]
.RB [ -b ]
.RB [ -c\ startcol ]
.RB [ -d\ tmpdir ]
.RB [ -e ]
.RB [ -f ]
.RB [ -h ]
.RB [ -i ]
.RB [ -L ]
.RB [ -l\ linenum ]
.RB [ -m\ maxreplines ]
.RB [ -n ]
.RB [ -P ]
.RB [ -p ]
.RB [ -r|-R ]
.RB [ -s ]
.RB [ -T ]
.RB [ -t\ maxtimes ]
.RB [ -u\ backupsuffix ]
.RB [ -v ]
.RB [ -w ]
.RB [ -x\ suffix [-x...] ]
.RB [ -z ]
.RB old_str
.RB new_str
.RB [ -a\ old_str\ new_str ]
.RI [ filename\ [filename...] ]
.SH DESCRIPTION
.I replace
performs (by default) a case-insensitive search for the "word" old_str in each
text line and replaces it with new_str, which is (by default) case-adjusted to
match the case of the particular occurrence of old_str in the original line.
A "word" is defined to be a string that is preceded AND followed by a
non\-alphanumeric character (start and end of lines count as such a character)
in the original text line. The original versions of any modified files are
(by default) saved with a ".cln" extension.
.PP
If no filenames are specified, then
standard input and output are used for the replacements unless -r or -R is
additionally specified
(which will cause the current directory to be recursed for files instead).
If "filenames" are specified with -r or -R, then they can either be files
or the names of directories that will be recursed for files.
If possible, modified files retain their ownership, group and permissions.
.SH OPTIONS
The following options are recognised:
.PP
.RS
.TP
.B  \-?
Display a syntax usage message.
.TP
.B  \-A
By default, replace uses the first 256 bytes of each file specified to
automatically determine if the file is text or binary. By specifying this
option, this auto-detection is turned off and replace assumes that all
supplied files are text files. It is recommended that you avoid this
option unless absolutely necessary due to the high risk of corruption
of binary files that may be accidentally specified. Because of its dangerous
nature, -A is now a deprecated option and may change its behaviour or be
removed in a future release.
.TP
.B  \-a old_str new_str
Allows extra replacement strings to be specified in strict left-to-right order.
The maximum number of replacement pairs is 255. This is equivalent to sed's
multiple -e parameters of course, so please note that the whole line is
modified by the first replacement pair before it is rescanned for subsequent
pairs. If -b or -h has been specified or the auto-detect code indicates
the file is a binary, the binary input data is read in 16K
chunks and the binary replacement pairs are applied in turn to each chunk
(chunk-straddling is catered for of course).
.TP
.B  \-b
Turns off file type auto-detection and forces replace to assume that all
supplied files are binary files (i.e. the inverse of the -A option).
The replacement string (new_str) must be less than or equal in
length to the original string (old_str) and will not be zero-terminated
unless -z is also supplied or padded with spaces unless -P or -p is used.
Note that all binary replacements do not check if old_str is a "word"
i.e. the -s option is enabled.
Neither old_str nor new_str can be zero length (i.e. "") when -b is used.
.PP
Note that -b turns on initial checks about the suitability of the old and
new strings for use in binary replacements (namely that old_str must not be
zero length and new_str must be less than or equal to the length of old_str).
Without the -b (or -h) option, such unsuitable strings are allowed, but will
be silently ignored if the auto-detect code finds a binary file and attempts
string replacements on it.
.TP
.B  \-c startcol
Starts the string replacement from column 'linenum' onwards rather than the
first column onwards. Ignored if -b or -h is also specified or if the
auto-detect code indicates the file is a binary.
.TP
.B  \-d tempdir
Specifies the temporary directory used for storing modified files. If the
parameter is not supplied, then the environmental variable TMPDIR is read
and used instead. If TMPDIR is also not set, then /tmp will be used.
.TP
.B  \-e
Make the search case-sensitive. new_str exactly replaces old_str with no
case\-adjustment to new_str.
.TP
.B  \-f
Forces the update of files without any backup (which typically has a .cln suffix).
.TP
.B  \-h
The original and replacement strings are normally ASCII strings, but -h
switches their format to be hex strings of binary data. It also activates
case-sensitive binary replacement mode (i.e. it switches on the -b and -e
options). If you wanted, for example, to replace all occurrences of the byte 0
with the byte 255 instead, you'd specify "-h 0 ff".
.TP
.B  \-i
When specified once, the user will be interactively prompted whether they
want the next file to have all of the appropriate strings replaced.
When specified twice ("-i -i"), the prompting switches to asking about every
single string replacement in each file (the appropriate unmodified line is
displayed to aid the user). As a safety precaution, verbose mode is also
switched on when -i is specified (to the level of the number of -i's
stated). Note that the input is read from /dev/tty
to avoid clashing with standard input and any prompts are, of course, sent
to standard error.
.TP
.B  \-L
Follows any soft-link specified on the command line to its linked-to
destination before performing any replacements.
If a soft-link eventually points to a directory and the -r or -R option
has been specified, then that directory will be recursed (but any
soft-links inside that directory will not be followed to avoid jumping
out of the directory tree). If -L isn't supplied, soft-links specified
on the command line will be skipped.
.TP
.B  \-l linenum
Starts the string replacement from line 'linenum' onwards rather than the first
line onwards. Ignored if -b or -h is also specified or if the file is
auto-detected as a binary.
.TP
.B  \-m maxreplines
Specifies the maximum number of lines that should have their strings replaced
in a file. The default value is infinity. Ignored if -b or -h is also
specified or if the file is auto-detected as a binary.
.TP
.B  \-n
Displays what strings would be replaced without actually replacing them.
It also, by necessity, switches on summary verbose mode (specify -nn to
switch on all replacements verbose mode).
.TP
.B  \-P
Pre-pads new_str with leading spaces if it is shorter than old_str.
.TP
.B  \-p
Pads new_str with trailing spaces if it is shorter than old_str.
.TP
.B  \-r or \-R
Recurses the filesystem from the current directory downwards (if no
"filenames" are specified), replacing strings in every file found.
Soft-links within the directory tree are ignored to avoid the recursion
jumping out of that tree.
The recursion can look for particular file suffixes only by the use of one or
more -x parameters. If "filenames" are specified along with -r or -R,
then they can either be files or the names of directories to recurse down for
files.  Note that files ending in the backup suffix (typically .cln) will be
skipped during the recursion.
.TP
.B  \-s
Relaxes the condition that old_str has to match a "word" i.e. it replaces
old_str even if it is a substring of a "word" or if it contains only
non-alphanumeric characters.
.TP
.B  \-T
Modified files retain the timestamps of the original unmodified files
if this option is specified. By default, modified files have their timestamp
set to the time the file was modified (i.e. now).
.TP
.B  \-t maxtimes
States the maximum number of times a string can be replaced in any single line
of a file. The default value is infinity. If -b or -h is also specified or
the file is auto-detected as a binary, the
maxtimes value applies to the whole binary input file.
.TP
.B  \-u backupsuffix
Changes the backup suffix of the unmodified file from the default of \.cln
to the specified backupsuffix value.
.TP
.B  \-v
Increments (switches on) verbose mode. If specified once, only a brief per-file
summary is reported to stderr of the files that had replacements performed on them.
If specified twice (i.e. -vv), all individual replacements are additionally reported 
to stderr and an overall summary is also displayed. If -v isn't specified at
all (and -i or -n aren't used), then only warnings and errors are displayed
to stderr (in other words, replace operates silently by default if there are
no problems).
.TP
.B  \-w
Recursively replaces strings in all Web-related source files
in the current directory tree. Equivalent to
"-r -x .html -x .htm -x .asp -x .js -x .css -x .xml -x .xhtml -x .shtml
-x .jsp -x .php -x .php3 -x .php4 -x .pl".
.TP
.B  \-x suffix
Specifies the suffix of the files to be case-insensitively matched when -r
is used.  Multiple -x parameters can be utilised with the appropriate "or"
meaning as you'd expect. The maximum number of -x parameters is 255.
.TP
.B  \-z
Zero-terminate any binary replacement string (new_str) specified, which is
useful if you are replacing a zero-terminated string (e.g. a pathname)
with a smaller string in an executable. If this option is used, -b must
also be supplied (note that using -h or -p will cause -z to be ignored)
and also new_str must be shorter than old_str in order to cater for the zero
terminator.
.SH RETURN VALUE
If replace has no non-fatal errors, it will return 0, otherwise it returns
1. Fatal errors can include bad command line options or parameters,
failure to allocate memory for replacement buffers, failure to read from
the terminal when expecting user input and exiting replace via the
"q" option when being prompted for input during "replace -i".
.SH EXAMPLES
To make an executable use /usr/lib/libc.1 instead of /usr/lib/libc.2 (note that
this is "risky" and may lead to run-time problems for the executable, but it's
one way to try to run an HP-UX 11.00 binary on HP-UX 10.20 !):
.PP
replace -bv /usr/lib/libc.2 /usr/lib/libc.1 filename
.PP
To make a file one long line (i.e. replace all line feeds with spaces):
.PP
replace -h a 20 filename
.PP
To recursively add the attribute bgcolor="#ffffff" to all <body> tags
(regardless of case) in all Web-related source documents in the current
directory tree downwards (but retaining the original case of the <body> tag):
.PP
replace -fswv "<body" "<body bgcolor=\\"#ffffff\\""
.SH TIPS
If the old or new string or the filenames you are specifying contain
a space or any special shell character (e.g. brackets, pipe, backquote
and so on), use single or double quotes to surround the whole string.
If you want to remove a string from a file, specify "" as the new string
to replace the old one.  If the old or new string begins with a minus
("-"), you will have to stop command line argument parsing with --
(e.g. replace -sev -- -old new...).
.SH WARNINGS
Replacing strings in binary executables, object files or libraries can lead to
permanent corruption of said files if you perform the replacements
in a haphazard manner. You are strongly advised to keep a backup copy of any
such binary files you intend to replace strings in.
.PP
Do not use the -A option (which is now officially deprecated)
unless you are completely convinced that all files
specified are text files. You risk severe corruption of any binaries specified
otherwise.
.SH SEE ALSO BUT I WOULDN'T BOTHER
sed(1)
.SH AUTHOR
Richard K. Lloyd	<replace@richardlloyd.org.uk>


CHANGES
This is a list of all the changes to replace since 1.00 was released:

replace 2.24 (October 2004)
============

* Corrected a comparison of two size_t's that were underflowing and causing
  a crash due to unallocated memory being accessed during string replacements
  in text files.
* One of the tests in the new test suite assumed that the user's PATH contained
  ".", which often isn't the case (particularly for root). This has been fixed.
* The "-A" option is deprecated in this release. It will either change its
  functionality (unlikely) or be removed (more likely) in a future release
  and hence should not be used any more. The reason for this is that it's
  too "dangerous" (can potentially damage binary files) and the current code
  isn't easily changed to make it less risky.

replace 2.23 (September 2004)
============

* Added a new -d parameter to specify which (temporary) directory
  modified files are written to (falls back to $TMPDIR or /tmp if -d
  isn't supplied).
* If a text file has only one line and that line has no linefeed (\n)
  terminator, then that single line is now correctly read in and parsed
  (it used to be incorrectly ignored in previous releases).
* If a text file has a string replaced in a line which has no linefeed
  terminator (by definition, this must be the last line in the file of course),
  then a linefeed is no longer added to the end of that replaced line.
* Avoid crash if -A is supplied and a binary file containing null bytes (ASCII
  code 0) is specified. This will *still* corrupt the output file, so a
  second warning about the -A option (i.e. don't use it unless you're
  certain all files are text files) was added to the man page as well.
* If fdopen() fails during the creation of a temporary file, that file
  is now correctly closed and deleted.
* Keep a global track of the name of the current temporary file (or ""
  to mean no temporary file active).
* The tidy_up routine (run on program exit) will now remove the current
  temporary file if one is active.
* Trap signals (e.g. SIGINT) and run the tidy_up routine. This finally means
  that, unlike "rival" string replacement programs, CTRL-C will no longer
  leave a temporary file behind.
* A new test suite has been written - "gmake test" now runs a shell script
  that puts the built replace binary through its paces. A more verbose
  version can be run with "gmake verbosetest".
* Ran the code through "splint" (Linux lint) and fixed the warnings.
* Updated FEATURES file to allow for GNU sed 4.X (a couple of marks extra
  for that compared to 3.X) and also added MySQL's replace (yes, there's a
  binary name-clash there with this program and I've told MySQL about that...).

replace 2.22 (May 2004)
============

* Support for signed long long added (including use of %lld for output of
  such 8-byte integers) - falls back to signed long if signed long long
  isn't available.
* Double-quoted source and destination files if mv binary used to rename
  files in the rename_files() routine. This should allow files with spaces
  in their name to be correctly renamed.
* Compiled with flags to support 64-bit file offsets/large files.
* Documented -P option in "replace -?" output.
* Fixed old and new string comparison when -h is used - memcmp() instead
  of strcmp() is needed of course because the hex strings can contain
  binary 0 in the middle of them, which thwarts strcmp().
* Minor changes to fix +w1 warnings with HP-UX 11.23's paranoid HP ANSI C
  compiler (most +w1 warnings now come from HP-UX's own system header
  files...<sigh>).
* Added note about shell special characters such as brackets in the man
  page (i.e. you need to single or double quote params if they include
  any special char).
* Confirmed that replace builds with a K&R-only C compiler ("cc_bundled"
  that comes as standard with HP-UX 11.00) as well as an ANSI C compiler.

replace 2.21 (March 2004)
============

* Fixed incorrect assumption that a char was an unsigned byte (it's
  actually a signed byte) - this was causing binary replacements to crash
  on occasions.
* Switched to using mkstemp() instead of tmpnam().
* Added .xhtml and .shtml extensions to those picked up by the -w flag.
* Makefile now requires GNU make and will auto-detect Linux or HP-UX flags
  as necessary.

replace 2.20 (June 2001)
============

* New auto.c source file added the auto-detection of binary files (turned
  on by default). At long last, a mixture of binary and text files can be
  supplied on the command line and replace will use the first 256 bytes
  of each file to determine if the file is text or binary automatically.
  Note that other string replacement programs such as sed or rpl actually
  corrupt binaries when they are run with different-sized old and new
  strings (by shifting all the data, making the binary fail to run),
  whereas the replace command avoids this.
* New FEATURES file to provide a feature-by-feature comparison chart between
  sed, rpl and replace. Guess who wins ? :-)
* Added new "-A" option to force replace to think all supplied files are
  text files (i.e. it turns off the new auto-detection code and reverts to
  the default text file replacement mode of previous releases). It's probably
  not a good idea to use this option, because there is a risk of corrupting
  any binaries that are specified.
* Added new "-P" option to pre-pad any new replacement string with leading
  spaces (as opposed to -p, which pads with trailing spaces).
* Verbose mode (-v) now has two levels - a single -v only displays a summary
  about files that had replacements performed on them. Two -v's (e.g. -vv)
  reverts to the older behaviour of showing all the replacements that take
  place and providing an overall summary. This means that replacing
  a large number of files (especially with -r) and using a single -v won't
  swamp the screen with output. -i and -n now also "double up" and increment

replace.c
/* replace.c - The ideal program for sed-haters (i.e. everyone).
   (C) Richard K. Lloyd 2001-2004

   Provides a simple way of replacing strings in lines without resorting to
   the ludicrous gibberish of sed, whose syntax was obviously dreamt up by
   someone on LSD !

   Syntax:
   replace [qualifiers] <oldstr> <newstr> [-a <oldstr> <newstr> [-a...]]
           [filename...]

   What "replace" does by default:

   It performs a case-insensitive search for the "word" <oldstr> in each text
   line and replaces it with <newstr>, which is case-adjusted to match the
   case of the particular occurrence of <oldstr> in the original line. A "word"
   is defined to be a string that is preceded AND followed by a non-
   alphanumeric character (start and end of lines count as such a character)
   in the original text line. Multiple strings can be replaced in order using
   the "-a" qualifier.

   The qualifiers modify this behaviour appropriately (try "replace -?"). */
  
#define Extern extern
#include "replace.h"

/* Routines only called from this source file and nowhere else */
extern int binary_io(P2(FILE *,FILE *)),text_io(P2(FILE *,FILE *)),
           is_binary(P(FILE *));
extern void init_vars(P(void));
extern void get_options(P2(int,char **));

/* Need to track original pwd if following soft links */
static char curpwd[BUFSIZ];

#ifdef __STDC__
int ask_user(int pval,char *askname)
#else
int ask_user(pval,askname)
int pval;
char *askname;
#endif
{
   /* Ask the user if they want to replace a string (or strings) */
   int invalid=(prompt==pval);
   while (invalid)
   {
      FILE *thand;
      char tbuff[BUFSIZ]; tbuff[0]='\0';
      if ((thand=fopen(TERMINAL_DEV,"r"))==(FILE *)NULL)
         leave("Can't open " TERMINAL_DEV " - aborted");
      (void)fprintf(stderr,"Replace %s (y/n/a/q) ? [y] ",askname);
      if (fgets(tbuff,(int)BUFSIZ,thand)==(char *)NULL)
         leave("Can't read terminal input - aborted");
      (void)fclose(thand);
      switch (tbuff[0])
      {
         case 'Y':
         case 'y':
         case '\n':
           (void)fprintf(stderr,"OK, will replace %s\n",askname);
           invalid=0;
           break;
         case 'N':
         case 'n':
           (void)fprintf(stderr,"Skipped replace of %s\n",askname);
           return(0);
         case 'A':
         case 'a':
           (void)fprintf(stderr,"OK, prompting has been turned off.\n");
           (void)fprintf(stderr,"All appropriate strings will be replaced in all specified files\n");
           prompt=0;
           return(1);
         case 'Q':
         case 'q':
           leave("Program quit at user's request");
           break;
         default:
           (void)fprintf(stderr,"Invalid response - please try again\n");
           break;
      }

   }
   return(1);
}

#ifdef __STDC__
void plural(char *str,LONG_LONG value)
#else
void plural(str,value)
char *str;
LONG_LONG value;
#endif
{
   /* Nicely format a numeric message. Note the kludgy check for <10 here -
      this is so that the switch value can be cast down to a long safely.
      Yep, dumb HP-UX 11.00 HP ANSI C compiler refuses to switch on a
      long long value <rolls eyes upwards> */
   if (value<10)
   switch ((long)value)
   {
      case 0: (void)fprintf(stderr,"no %ss",str); break;
      case 1: (void)fprintf(stderr,"one %s",str); break;
      case 2: (void)fprintf(stderr,"two %ss",str); break;
      case 3: (void)fprintf(stderr,"three %ss",str); break;
      case 4: (void)fprintf(stderr,"four %ss",str); break;
      case 5: (void)fprintf(stderr,"five %ss",str); break;
      case 6: (void)fprintf(stderr,"six %ss",str); break;
      case 7: (void)fprintf(stderr,"seven %ss",str); break;
      case 8: (void)fprintf(stderr,"eight %ss",str); break;
      default: (void)fprintf(stderr,"nine %ss",str); break;
   } 
   else (void)fprintf(stderr,LONG_LONG_FORMAT " %ss",value,str);
}

#ifdef __STDC__
static int read_lines(FILE *fd,FILE *fdout,LONG_LONG filesize,char *filename)
#else
static int read_lines(fd,fdout,filesize,filename)
FILE *fd,*fdout;
LONG_LONG filesize;
char *filename;
#endif
{
   /* Handle a whole file. Returns 0 = success, != 0 for failure */
   int gotfail,this_binary=is_binary(fd);
   if (verbose==2)
   {
      (void)fprintf(stderr,"Scanning %s %s",filetype,filename);
      if (filesize>=0)
      {
         (void)fprintf(stderr," (");
         plural("byte",filesize);
         (void)fprintf(stderr,")");
      }
      (void)fprintf(stderr,"...\n");
   }
   linenum=0; repcount=0; numfiles++;
   if (this_binary) gotfail=binary_io(fd,fdout);
   else gotfail=text_io(fd,fdout);
   if (verbose && !gotfail)
   {
      if (verbose==2 || repcount)
      {
         (void)fprintf(stderr,"The %s %s ",filetype,intitle);
         if (!this_binary)
         {
            (void)fprintf(stderr,"contained ");
            plural("line",(LONG_LONG)linenum);
            (void)fprintf(stderr," and\n");
         }
         if (fake) (void)fprintf(stderr,"would have ");
         (void)fprintf(stderr,"had ");
         if (hex) plural("binary hex string",(LONG_LONG)repcount);
         else plural("string",(LONG_LONG)repcount);
         (void)fprintf(stderr," replaced");
         if (!this_binary)
         {
            (void)fprintf(stderr," in ");
            plural("line",(LONG_LONG)linecount);
         }
         (void)fprintf(stderr,"\n");
      }
   }
   return(gotfail);
}

#ifdef __STDC__
static int rename_file(char *source,char *dest)
#else
static int rename_file(source,dest)
char *source,*dest;
#endif
{
    /* Rename file source to file dest.
       Returns 0 for failure, 1 for success */
    if (rename(source,dest))
    {
       /* Note that the rename function CANNOT rename across filing systems,
          but /bin/mv CAN, so might as well try /bin/mv if rename fails.
          This is a weakness of "rpl -t" - if you set TMPDIR=/tmp, then
          try to replace strings in a file not on the same disk as /tmp
          using "rpl -t", you can't do it ! */
       char command[BUFSIZ];
       (void)snprintf(command,BUFSIZ,"%s \"%s\" \"%s\" 2>/dev/null",RENAME_COMMAND,source,dest);
       if (system(command)) return(0);
    }
    return(1);
}

#ifdef __STDC__
static int update_file(char *oldname,char *newname,char *orgname)
#else
static int update_file(oldname,newname,orgname)
char *oldname,*newname,*orgname;
#endif
{
   /* File needs to be updated and possibly a backup version kept */
   struct stat finfo;
   (void)stat(orgname,&finfo);
   if (rename_file(oldname,newname))
   {
      updated=1;
      if (verbose)
      {
         struct stat fst;
         LONG_LONG new_file_size;
         (void)stat(newname,&fst);
         new_file_size=(LONG_LONG)fst.st_size;
         (void)fprintf(stderr,"Updated %s %s (",filetype,newname);
         plural("byte",new_file_size);
         if (force) (void)fprintf(stderr,")\n");
         else (void)fprintf(stderr," - backup with %s extension)\n",backupsuff);
      }
      if (chmod(newname,finfo.st_mode))
        (void)fprintf(stderr,"WARNING: Couldn't set permissions of %s %s\n",filetype,newname);
      if (chown(newname,finfo.st_uid,finfo.st_gid))
        (void)fprintf(stderr,"WARNING: Couldn't set ownership of %s %s\n",filetype,newname);
      if (retainstamp)
      {
         struct utimbuf oldtime;
         oldtime.actime=finfo.st_atime;
         oldtime.modtime=finfo.st_mtime;
         if (utime(newname,&oldtime))
            (void)fprintf(stderr,"WARNING: Couldn't retain timestamp of %s %s\n",filetype,newname);
      }
      return(1);
   }
   else return(0);
}

#ifdef __STDC__
static void replace_a_file(const char *thename)
#else
static void replace_a_file(thename)
char *thename;
#endif
{
   /* Replace strings in a file, writing the new version to a temporary
      file which is either then deleted (no replacements) or renamed onto
      the original (some replacements) */
   FILE *filedes,*fileout=(FILE *)NULL;
   char repmess[BUFSIZ];

   (void)strcpy(intitle,thename);
   (void)snprintf(repmess,BUFSIZ,"strings in %s",intitle);
   if (!ask_user(1,repmess)) return;
   if ((filedes=fopen(intitle,"r"))==(FILE *)NULL)
        (void)fprintf(stderr,"WARNING: Skipped %s - could not open\n",
                intitle);
   else
   {
      char tmpname[BUFSIZ];
      int tmphandle=0;
      (void)snprintf(tmpname,BUFSIZ,"%s/replace_tmp_XXXXXX",tmpdir);
      if (!fake && (tmphandle=mkstemp(tmpname))==-1)
      {
         (void)fprintf(stderr,"WARNING: Skipped %s - couldn't create temp file %s\n",intitle,tmpname);
         (void)fclose(filedes);
      }
      else
      { 
         if (!fake) set_temp_file(tmpname);
         updated=0;
         if (!fake && (fileout=fdopen(tmphandle,"r+"))==(FILE *)NULL)
         {
            (void)fprintf(stderr,
                    "WARNING: Skipped %s - couldn't associate stream with temp file %s\n",
                    intitle,tmpname);
            (void)fclose(filedes);
            (void)close(tmphandle);
         }
         else
         {
            int gotfail;
            struct stat fst;
            LONG_LONG cur_file_size;

            (void)stat(intitle,&fst);
            cur_file_size=(LONG_LONG)fst.st_size;
            gotfail=read_lines(filedes,fileout,cur_file_size,intitle);
            if (!fake) gotfail|=fclose(fileout);
            gotfail|=fclose(filedes);
            if (gotfail)
               (void)fprintf(stderr,"WARNING: Read/write/close failure when accessing %s - skipped\n",intitle);
            else
            if (repcount)
            {
               int goodone=1;
               if (!fake)
               {
                  if (force) goodone=update_file(tmpname,intitle,intitle);
                  else
                  {
                     char outtitle[BUFSIZ];
                     (void)strcpy(outtitle,intitle);
                     (void)strcat(outtitle,backupsuff);
                     if ((goodone=rename_file(intitle,outtitle)))
                        goodone=update_file(tmpname,intitle,outtitle);
                  }
               }
               if (goodone) { numfilereps++; numreps+=repcount; }
               else (void)fprintf(stderr,"WARNING: Failed to update %s %s\n",filetype,intitle);
            }
         }
         if (!fake && !updated) (void)unlink(tmpname);
         if (!fake) set_temp_file("");
      }
   }
}

#ifdef __STDC__
static int do_branch(const char *fname,const struct stat *flstat,int fint)
#else
static int do_branch(fname,flstat,fint)
char *fname;
struct stat *flstat;
int fint;
#endif
{
   /* Walk down a directory tree, replacing strings in appropriately
      matching files */

#ifdef __STDC__
   const
#endif
   char *tmpfname=fname;
   int suffmatch=0;
   size_t flen,sloop;
   /* The next line is just to make the filename neater */
   if (!strncmp(tmpfname,"./",2)) { tmpfname+=2; }
   flen=strlen(tmpfname);
   switch (fint)
   {
      case FTW_F:
         if (flen<sufflen ||
             strncmp(&tmpfname[flen-sufflen],backupsuff,sufflen))
         {
            if (!suffixes) replace_a_file(tmpfname);
            else
            if (minsuff<=flen)
            for (sloop=1;sloop<=suffixes && !suffmatch;sloop++)
            {
               char *sptr=sufflist[sloop];
               size_t sxlen=strlen(sptr);
               if (!strncasecmp(&tmpfname[flen-sxlen],sptr,sxlen))
                 replace_a_file(tmpfname);
            }
         }
      break;
   }
   return(0);
}

#ifdef __STDC__
static void recurse_down(char *recdir)
#else
static void recurse_down(recdir)
char *recdir;
#endif
{
   /* Recursive replacements requested */
   if (verbose) (void)fprintf(stderr,"Recursing down %s dir tree\n",recdir);
   if (ftw(recdir,do_branch,MAX_DIR_LEVELS))
     (void)fprintf(stderr,"WARNING: Failed to recurse down %s dir tree\n",recdir);
}

static void finish_it(P(void))
{
   /* All replacements/files done, so display a summary if in verbose mode 2 */
   if (verbose==2 && numfiles>1)
   {
      (void)fprintf(stderr,"Number of files scanned: %d\n",numfiles);
      (void)fprintf(stderr,"Number of files that ");
      if (fake) (void)fprintf(stderr,"would have ");
      (void)fprintf(stderr,"had strings replaced: %d\n",numfilereps);
      (void)fprintf(stderr,"Number of ");
      if (fake) (void)fprintf(stderr,"potential ");
      (void)fprintf(stderr,"string replacements in total: %d\n",numreps);
   }
   tidy_up();
}

#ifdef __STDC__
static char *dirname(char *pname)
#else
static char *dirname(pname)
char *pname;
#endif
{
   /* Get the directory name of the passed filename path */
   static char dname[BUFSIZ],*dptr;
   (void)strcpy(dname,pname);
   if (!strcmp(pname,"/")) return("/");
   if ((dptr=strrchr(dname,'/'))!=(char *)NULL)
   {
      *dptr='\0'; 
      return(dname);
   } else return(".");
}

#ifdef __STDC__
char *basename_path(char *pname)
#else
char *basename_path(pname)
char *pname;
#endif
{
   /* Get the leafname of the passed filename path */
   char *dptr;
   if (!strcmp(pname,"/")) return("/");
   if ((dptr=strrchr(pname,'/'))!=(char *)NULL)
     return(&dptr[1]); else return(pname);
}

#ifdef __STDC__
static void do_the_file(char *fname,int sofar)
#else
static void do_the_file(fname,sofar)
char *fname;
int sofar;
#endif
{
   /* Run appropriate function for the filename passed. Note that this
      routine can be recursive (if followsoftlinks is on) */
   struct stat finfo;
   if (!lstat(fname,&finfo))
   {
      if (S_ISDIR(finfo.st_mode))
      {
         if (recursive) recurse_down(fname);
         else (void)fprintf(stderr,"WARNING: Skipped %s directory - use -r for recursion\n",fname);
      }
      else
      if (S_ISLNK(finfo.st_mode))
      {
         if (followsoftlinks)
         {
            char linkbuf[BUFSIZ];
            if (readlink(fname,linkbuf,BUFSIZ)<0)
              (void)fprintf(stderr,"WARNING: Can't read %s soft-link - skipped\n",fname);
            else
            if (sofar==MAX_SOFT_LINKS)
              (void)fprintf(stderr,"WARNING: Soft-link loop (%s) detected - skipped\n",fname);
            else
            if (linkbuf[0]=='/')
            do_the_file(linkbuf,sofar+1);
            else
            {
               char linkdest[BUFSIZ],fdir[BUFSIZ];
               (void)strcpy(fdir,dirname(fname));
               (void)snprintf(linkdest,BUFSIZ,"%s/%s",fdir,dirname(linkbuf));
               if (chdir(linkdest) || getcwd(linkdest,BUFSIZ)==(char *)NULL)
               {
                  (void)fprintf(stderr,"WARNING: Soft-link (%s) to non-existent dir - skipped\n",fname);
                  (void)chdir(curpwd);
               }
               else
               {
                  (void)chdir(curpwd);
                  (void)strcat(linkdest,"/");
                  (void)strcat(linkdest,basename_path(linkbuf));
                  do_the_file(linkdest,sofar+1);
               }
            }   
         }
         else (void)fprintf(stderr,"WARNING: Soft-link (%s) detected - skipped\n",fname);
      }
      else
      if (S_ISREG(finfo.st_mode))
      {
         if (finfo.st_size) replace_a_file(fname);
         else
         (void)fprintf(stderr,"WARNING: %s is zero-length - skipped\n",fname);
      }
      else (void)fprintf(stderr,"WARNING: %s is not a file/dir - skipped\n",fname);
   }
   else (void)fprintf(stderr,"WARNING: No such file/dir (%s) - skipped\n",fname);
}

void check_tmp_dir_exists(P(void))
{
   /* Make sure tmpdir is actually a directory - it's fatal if it isn't,
      because we can't do any replacements in files without it */
   struct stat tmpdirinfo;
   if (stat(tmpdir,&tmpdirinfo) || !S_ISDIR(tmpdirinfo.st_mode))
      leave("Temporary directory not found");
}

#ifdef __STDC__
int main(int argc,char **argv)
#else
int main(argc,argv)
int argc;
char **argv;
#endif
{
   /* Parse command line options and then either read from stdin/send to
      stdout, recurse down or do replacements on specified files */
   init_vars();
   get_options(argc,argv);
   if (optind==argc || !strcmp(argv[optind],"-"))
   {
      if (recursive)
      {
         check_tmp_dir_exists();
         recurse_down(".");
      }
      else
      {
         if (ask_user(1,"strings in standard input"))
         {
            (void)strcpy(intitle,"<stdin>");
            if (read_lines(stdin,stdout,(LONG_LONG)-1,intitle))
              (void)fprintf(stderr,"WARNING: Read/write error during stdin/stdout operations - aborted\n");
         }
      }
   }
   else
   {
      if (followsoftlinks)
      {
         if (getcwd(curpwd,BUFSIZ)==(char *)NULL)
            leave("Can't determine current working directory"); 
      }
      check_tmp_dir_exists();
      for (;optind<argc;optind++) do_the_file(argv[optind],0);
   }
   finish_it();
   return(0);
}
